plugins/inputs/solr/solr.go

```diff
@@ -0,0 +1,424 @@
+			"class_name":                  metrics.Class,
+			"adds":                        metrics.Stats.Adds,
+			"autocommit_max_docs":         metrics.Stats.AutocommitMaxDocs,
+			"autocommit_max_time":         autoCommitMaxTime,
+			"autocommits":                 metrics.Stats.Autocommits,
+			"commits":                     metrics.Stats.Commits,
+			"cumulative_adds":             metrics.Stats.CumulativeAdds,
+			"cumulative_deletes_by_id":    metrics.Stats.CumulativeDeletesByID,
+			"cumulative_deletes_by_query": metrics.Stats.CumulativeDeletesByQuery,
+			"cumulative_errors":           metrics.Stats.CumulativeErrors,
+			"deletes_by_id":               metrics.Stats.DeletesByID,
+			"deletes_by_query":            metrics.Stats.DeletesByQuery,
+			"docs_pending":                metrics.Stats.DocsPending,
+			"errors":                      metrics.Stats.Errors,
+			"expunge_deletes":             metrics.Stats.ExpungeDeletes,
+			"optimizes":                   metrics.Stats.Optimizes,
+			"rollbacks":                   metrics.Stats.Rollbacks,
+			"soft_autocommits":            metrics.Stats.SoftAutocommits,
+		}
+		acc.AddFields(
+			fmt.Sprintf("solr_%s", strings.ToLower(category)),
+			coreFields,
+			map[string]string{
+				"core":    core,
+				"handler": name},
+			measurementTime,
+		)
+	}
+	return nil
+}
+
+func gatherCacheMetrics(mbeansJSON json.RawMessage, core, category string, acc telegraf.Accumulator) error {
+	var coreMetrics map[string]Cache
+
+	measurementTime := time.Now()
+	if err := json.Unmarshal(mbeansJSON, &coreMetrics); err != nil {
+		return err
+	}
+	for name, metrics := range coreMetrics {
+		coreFields := map[string]interface{}{
+			"class_name":           metrics.Class,
+			"cumulative_evictions": metrics.Stats.CumulativeEvictions,
+			"cumulative_hitratio":  metrics.Stats.CumulativeHitratio,
+			"cumulative_hits":      metrics.Stats.CumulativeHits,
+			"cumulative_inserts":   metrics.Stats.CumulativeInserts,
+			"cumulative_lookups":   metrics.Stats.CumulativeLookups,
+			"evictions":            metrics.Stats.Evictions,
+			"hitratio":             metrics.Stats.Hitratio,
+			"hits":                 metrics.Stats.Hits,
+			"inserts":              metrics.Stats.Inserts,
+			"lookups":              metrics.Stats.Lookups,
+			"size":                 metrics.Stats.Size,
+			"warmup_time":          metrics.Stats.WarmupTime,
+		}
+		acc.AddFields(
+			fmt.Sprintf("solr_%s", strings.ToLower(category)),
+			coreFields,
+			map[string]string{
+				"core":    core,
+				"handler": name},
+			measurementTime,
+		)
+	}
+	return nil
+}
+
+func (s *Solr) gatherCoreStats(url string, core string, acc telegraf.Accumulator) error {
+
+	for _, category := range []string{"CORE", "QUERYHANDLER", "UPDATEHANDLER", "CACHE"} {
+		metrics := &Metrics{}
+		if err := s.gatherData(fmt.Sprintf("%s&cat=%s", url, category), metrics); err != nil {
+			acc.AddError(fmt.Errorf("JSON fetch error: %s", err))
+		}
+		switch category {
+		case "CORE":
```

Comment:
This feels rather odd, creating a static slice, iterating it, and then switching on which one you're currently iterating on. What about a function you pass the appropriate gather func to. For example:

```
err = s.gatherCategory(url, core, "CORE", gatherCoreMetrics, acc)
err = s.gatherCategory(url, core, "QUERYHANDLER", gatherQueryHandlerMetrics, acc)
...
```